src/lib/device-online-state/index.ts

```diff
@@ -0,0 +1,282 @@
+import { DEFAULT_SUPERVISOR_POLL_INTERVAL } from '../env-vars';
+import { noop } from 'lodash';
+import { resinApi, root } from '../../platform';
+import { captureException } from '../../platform/errors';
+import {
+	createPromisifedRedisClient,
+	PromisifedRedisClient,
+} from '../redis-promise';
+import * as RedisSMQ from 'rsmq';
+import { REDIS_HOST, REDIS_PORT, API_HEARTBEAT_STATE_TIMEOUT } from '../config';
+
+export class PollIntervalNotFoundError extends Error {}
+export const getPollIntervalForDevice = (uuid: string): Bluebird<number> => {
+	return resinApi
+		.get({
+			resource: 'device_config_variable',
+			passthrough: {
+				req: root,
+			},
+			options: {
+				$select: 'value',
+				$expand: {
+					device: {
+						$filter: { uuid },
+					},
+				},
+				$filter: {
+					name: 'RESIN_SUPERVISOR_POLL_INTERVAL',
+				},
+			},
+		})
+		.then(([pollInterval]: Array<{ value: number }>) => {
+			if (!pollInterval) {
+				throw new PollIntervalNotFoundError();
+			}
+
+			return pollInterval.value;
+		})
+		.catch(PollIntervalNotFoundError, () => {
+			// query to determine the poll interval for this device...
+			return resinApi
+				.get({
+					resource: 'application_config_variable',
+					passthrough: {
+						req: root,
+					},
+					options: {
+						$select: 'value',
+						$top: 1,
+						$filter: {
+							application: {
+								$any: {
+									$alias: 'a',
+									$expr: {
+										a: {
+											owns__device: {
+												$any: {
+													$alias: 'd',
+													$expr: {
+														d: {
+															uuid: uuid,
+														},
+													},
+												},
+											},
+										},
+									},
+								},
+							},
+							name: 'RESIN_SUPERVISOR_POLL_INTERVAL',
+						},
+					},
+				})
+				.then(([pollInterval]: Array<{ value: number }>) => {
+					if (pollInterval) {
+						return pollInterval.value;
+					}
+
+					return DEFAULT_SUPERVISOR_POLL_INTERVAL;
+				});
+		})
+		.then(pollInterval => {
+			return Math.max(pollInterval, DEFAULT_SUPERVISOR_POLL_INTERVAL);
+		});
+};
+
+export enum DeviceOnlineStates {
+	Unknown = -2,
+	Timeout = -1,
+	Offline = 0,
+	Online = 1,
+}
+
+class NoMessagesToConsumeSignal extends Error {}
+
+class DeviceOnlineStateManager extends EventEmitter {
+	private static readonly REDIS_NAMESPACE = 'device-online-state';
+	private static readonly EXPIRED_QUEUE = 'expired';
+
+	rsmq: RedisSMQ;
+	redis: PromisifedRedisClient;
+
+	constructor() {
+		super();
+
+		// create a new Redis client...
+		this.redis = createPromisifedRedisClient({
+			host: REDIS_HOST,
+			port: REDIS_PORT,
+		});
+
+		// initialise the RedisSMQ object using our Redis client...
+		this.rsmq = new RedisSMQ({
+			client: this.redis,
+			ns: DeviceOnlineStateManager.REDIS_NAMESPACE,
+		});
+
+		// create the RedisMQ queue and start consuming messages...
+		this.rsmq
+			.createQueueAsync({ qname: DeviceOnlineStateManager.EXPIRED_QUEUE })
+			.catch((err: Error) => {
```

Comment:
You should be able to remove the type annotation, there may be an issue with typings if not.